packages/vite/src/node/plugins/css.ts

```diff
@@ -556,16 +563,19 @@ export function cssPostPlugin(config: ResolvedConfig): Plugin {
           })
           chunk.viteMetadata.importedCss.add(this.getFileName(fileHandle))
         } else if (!config.build.ssr) {
-          // legacy build and inline css
-
-          // __VITE_ASSET__ and __VITE_PUBLIC_ASSET__ urls are processed by
-          // the vite:asset plugin, don't call resolveAssetUrlsInCss here
+          // legacy build, inline css
+          // the entry css(style.css) will be inlined in the html template
+          // so don't need to generate the css insert script
+          if (chunk.isEntry) {
+            return { code }
+          }
+          chunkCSS = resolveAssetUrlsInCss(chunkCSS, chunk.name, true)
```

Comment:
If this branch is only hit in legacy mode, then even if the `vite:asset` plugin were reached the code there would be wrong as it assumes we are working in an ES module (so it uses `import.meta.url` for example). So resolving here seems ok. But the current code will break if the assets and the HTML are deployed separately, so this will break if the user uses `renderBuiltUrl` with plugin legacy. I think in the future we may need to call `toOutputFilePathInString` instead of `toOutputFilePathInCss`, see the functions here https://github.com/vitejs/vite/blob/10757b84009b27744b0645b75e7951bbbac6a928/packages/vite/src/node/build.ts#L839
But `renderBuiltUrl` gives as hostType `'html' | 'css' | 'js'`. So we may need a new hostType if we are doing this. Maybe `'js-nomodule'`?
This sounds like a new feature though instead of a fix, and I don't know how much it would be used. @danielroe for Nuxt, I imagine you are using vite plugin legacy, no? Did you experienced issues with inlined CSS?

So for the current fix, maybe `toOutputFilePathInCss` is fine as a temporal measurement. I think that `self.location` for relative base will only work if the HTML and the assets are not deployed separatedly, maybe this is the only thing we can do though as we don't have `import.meta.url`.
I'll think a bit more, leaning towards merging this and then checking if we can implement better support for `renderBuiltUrl` and legacy + inlined CSS

Reply:
mmm... ok, scratch my prev message. Why is `vite:asset` not being hit for these modules? we don't need `js-nomodule`, it should be always ESM inside the vite pipeline and plugin-legacy should be transforming `import.meta.url` to something that works in legacy browsers.
I think what we need is for `vite:asset` to transform this modules or to call its code if somehow it is being short-circuited @poyoho

Reply:
> Why is `vite:asset` not being hit for these modules?

This is because `vite:asset` is using `renderChunk` hook and `vite:css-post` is using `generateBundle` hook.

> it should be always ESM inside the vite pipeline and plugin-legacy should be transforming import.meta.url to something that works in legacy browsers.

I wrote a explaination about this at https://github.com/vitejs/vite/pull/9381.

Reply:
Sorry, I didn't see #9381 before! awesome you drilled into that.

>> Why is `vite:asset` not being hit for these modules?

> This is because `vite:asset` is using `renderChunk` hook and `vite:css-post` is using `generateBundle` hook.

`vite:css-post` is using both `renderChunk` and `generateBundle`, and this branch we are discussing is on `renderChunk`. So this shouldn't be the reason.

Reply:
> > > Why is `vite:asset` not being hit for these modules?
> >
> > This is because `vite:asset` is using `renderChunk` hook and `vite:css-post` is using `generateBundle` hook.
> 
> `vite:css-post` is using both `renderChunk` and `generateBundle`, and this branch we are discussing is on `renderChunk`. So this shouldn't be the reason.

My bad. I was overlooking that. It's because `vite:asset` comes before `vite:css-post`.
https://github.com/vitejs/vite/blob/6d952252c7a3b92b2f82a7c78c230e0a152fd2a6/packages/vite/src/node/plugins/index.ts#L83-L87

Reply:
I see. We have a few options:

- we could break the asset plugin into `vite:asset` and `vite:asset-post` (and move the `renderChunk` there). But this is a big change because `normalPlugins` may now assume that the `__VITE_ASSET` are resolved in their `renderChunk`. So we should only do this in a major, and I'm not sure is ideal.
- we could move the `renderChunk` from `css-post` to `css`, or maybe not all of it, only the resolving of `__VITE_ASSET` part. Again, this may introduce subtle regressions in the ecosystem and will need a major. Only thinking on this to try to make the resolution happen for both JS and CSS in the same place (or both before normal plugins, or both afterwards, not we have a mix)
- For this fix in particular, so we can include it in a patch, I think the direction of this PR is correct, we should directly resolve in `css-post`. But we should use the same logic we have in `vite:asset` so `renderBuiltUrl` is respected. @sapphi-red @poyoho what do you think if we extract what we do in `vite:asset` `renderChunk` hook in a helper that we can call here too?

Reply:
IMO the best choice is the first one. But I agree we shouldn't do that in minor or patch.
To fix this in a patch, I am for this PR's approach, too. Extracting in a helper function sounds good to me.

Reply:
I like first. It make `__VITE_ASSET` in one space.

Reply:
@poyoho let me know if you are going to work on implementing the third approach so we can patch plugin-legacy in 3.0. If you are busy, I could implement it later this week.

Reply:
maybe I do it in tomorrow night